### ISSUE TITLE
Subprocess should emit error if executable path doesn't exist in Windows

### ISSUE THREAD

**Issue by lynxtaa** [sindresorhus/execa#469]
Returned from `execa` subprocess should emit 'error' event if executable path doesn't exist in Windows. Just like `spawn` and `execFile` in `child_process`

Consider this code:

```js
const { spawn } = require('child_process')
const execa = require('execa')

const p1 = spawn('wrong-path')
p1.on('error', err => {
	console.log(err) // logs ENONENT error
})

const p2 = execa('wrong-path')
p2.on('error', err => {
	console.log(err) // never called in Windows, but called in Linux
})
```

**Comment by ehmicky** [comment]
Hi @lynxtaa,

Thanks for reaching out.
Which version of `execa` are you using? I cannot reproduce this behavior.

This is an example with running your example on my machine on Windows. Note: I replaced the second `wrong-path` by `wrong-path-2` to differentiate them.

![windows](https://user-images.githubusercontent.com/8136211/121517844-18c79880-c9f0-11eb-951a-37f0637e38af.png)

A potential issue you might be experiencing is if you forgot to add a `.catch()` handler to the return value of `execa`. The return value is both  a child process and a promise. On errors, that promise is rejected. Unhandled promise rejections make the process exit by default on Node.js, which might prevent further code (like the `error` event handler) from running.

Aside from this, `execa` actually calls `spawn()` under the hood and just return the child process as is, so there should not be any difference of behavior.

**Comment by lynxtaa** [comment]
@ehmicky I'm running Node@16.3.0 and execa@5.1.1.

> A potential issue you might be experiencing is if you forgot to add a `.catch()` handler to the return value of `execa`. The return value is both a child process and a promise. On errors, that promise is rejected. Unhandled promise rejections make the process exit by default on Node.js, which might prevent further code (like the `error` event handler) from running.

I've tried running `node  --unhandled-rejections=warn t.js` -- same issue

**Comment by ehmicky** [comment]
Is `t.js` contained anymore more than the sample above?
Considering I get a different behavior when running it myself on Windows (in both `cmd.exe` and Bash), I am curious which differences of environment might explain that we get different results? Is there any additional reason you think might explain those differences?

**Comment by lynxtaa** [comment]
> Does t.js contain any additional code beside the sample above?

No, that's all. I've reproduced this problem on three different Windows 10 machines running Node 16.3.0 and Node 14.17.0. In WSL in works okay. To be sure it has time to handle an error, I'm running this code:
```js
const execa = require('execa')

const p = execa('wrong-path')
p.on('error', err => {
  console.log(err) // never called in Windows, but called in Linux
})

setTimeout(() => {}, 1000)
```

Still no luck

**Comment by ehmicky** [comment]
Would it be possible to create a GitHub repository to reproduce this behavior, with a GitHub action running on Windows?
This could help us figure out what is happening. Thanks!

**Comment by lynxtaa** [comment]
Done https://github.com/lynxtaa/execa-issue 

![scr](https://user-images.githubusercontent.com/22496946/121526508-1b32ee00-ca02-11eb-9e82-c3ca8a224bbe.png)

**Comment by ehmicky** [comment]
Thanks a lot @lynxtaa, that's perfect :+1: 
On your machine, are you using Powershell? If so, could you try using `cmd.exe` instead?

**Comment by lynxtaa** [comment]
I'm using `cmd.exe`

**Comment by lynxtaa** [comment]
Running with cmd in CI. Same result:

![asd](https://user-images.githubusercontent.com/22496946/121528050-c6907280-ca03-11eb-83b1-72fc0d64a19c.png)

**Comment by lynxtaa** [comment]
After further investigation I think the issue is in `cross-spawn` arguments parsing. Since then binary `wrong-path` doesn't exist `cross-spawn` will interpret it as a shell command. After adding `console.log(parsed)` here https://github.com/sindresorhus/execa/blob/9216ec8035f55a3ddcbf07de8667f9d9d5c40c84/index.js#L83 and running on Windows:
```sh
{
  file: 'C:\\WINDOWS\\system32\\cmd.exe',
  args: [ '/d', '/s', '/c', '"wrong-path"' ],      
  options: {
    maxBuffer: 100000000,
    buffer: true,
    stripFinalNewline: true,
    extendEnv: true,
    preferLocal: false,
```

Running on Linux:
```sh
{
  file: 'wrong-path',
  args: [],
  options: {
    maxBuffer: 100000000,
    buffer: true,
    stripFinalNewline: true,
    extendEnv: true,
    preferLocal: false,
```

**Comment by lynxtaa** [comment]
Related https://github.com/moxystudio/node-cross-spawn/issues/16

**Comment by ehmicky** [comment]
Thanks for digging into this @lynxtaa!

The actual issue seems to be the following: Node.js detects `ENOENT` on Windows by translating many Windows system error codes to `ENOENT` in `libuv` ([link to code](https://github.com/libuv/libuv/blob/46451737e6174281317cabdf05c9ba1434c359d6/src/win/error.c#L133)). When this happens, an `error` event is emitted on the `child_process` ([link to code](https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/child_process.js#L393)). However, `node-cross-spawn` sometimes wraps the command with `cmd.exe` in order to support shebangs. This makes the above Node.js behavior not work anymore. This is why you're experiencing different behaviors between Execa and `child_process`.

`libuv` has access to the Windows API system error codes. However, we cannot use those error codes userland (since they are abstracted away by `libuv`). Unfortunately, the solution implemented in `node-cross-spawn`, which relies on the exit code being `1` is not reliable as described in https://github.com/moxystudio/node-cross-spawn/issues/104, which is why Execa avoids it.

**Comment by lynxtaa** [comment]
@ehmicky Thanks you for this detailed explanation.

There are cases when we are running a subprocess for a very long time and awaiting a promise is undesirable. To detect a startup failure with `execa` on Windows I may use something like this for now:
```js
await Promise.race([
  subprocess,
  new Promise(resolve => setTimeout(resolve, 1000))
])
``` 

But I still think some kind of error should be emitted when using `on('error', ...)`, maybe not `ENOENT` but _exiting with code 1 error_  which I can see, when using `await execa(...)`. What do you think?

**Comment by ehmicky** [comment]
Would `subprocess.catch(error => { ... })` work? It would behave identically as the error handling behavior you're describing for non-existing executable paths.

**Comment by lynxtaa** [comment]
I'm very grateful for this detailed write-up. Thank you, @ehmicky ❤️ !

Yeah, I've wanted to use `execa` to run GUI application and only check if it launched successfully (e.g. path to a binary file exists) and don't wait for it to exit since it may take a long time.

Previously I've used `child_process.spawn` with `on('error')` event listener but now I've discovered that `spawn` doesn't emit this event for startup failures (e.g. subprocess exits with a `code 1`). So using `execa` with `Promise.race` seems like a more reliable solution and I'm pretty happy about it
